File: src/hypeedge/market_data/book.py
"""In-memory L2 order book manager."""

from __future__ import annotations

from collections.abc import Iterator
from datetime import UTC, datetime

import structlog

from hypeedge.core.models import L2BookSnapshot, L2Level
from hypeedge.core.types import Price, Size, Symbol, Timestamp

logger = structlog.get_logger(__name__)


class OrderBook:
    """Maintains an in-memory L2 order book for a single symbol.

    Updated from WebSocket feed, readable by strategies with zero latency.
    Thread-safe for single-event-loop use (asyncio).
    """

    def __init__(self, symbol: Symbol, depth: int = 20) -> None:
        self.symbol = symbol
        self._depth = depth
        self._bids: list[L2Level] = []  # Sorted: best (highest) first
        self._asks: list[L2Level] = []  # Sorted: best (lowest) first
        self._last_update_ts: Timestamp | None = None
        self._version = 0
        self._snapshot: L2BookSnapshot | None = None

    def update(
        self,
        bids: list[tuple[float, float]],
        asks: list[tuple[float, float]],
        ts: Timestamp,
        *,
        received_at: datetime | None = None,
        connection_generation: int = 0,
    ) -> L2BookSnapshot:
        """Update the book with a full snapshot from WebSocket.

        Args:
            bids: List of (price, size) tuples, best first
            asks: List of (price, size) tuples, best first
            ts: Exchange timestamp
            received_at: Local time at the WebSocket receive boundary
            connection_generation: Monotonic WebSocket reconnection generation
        """
        self._bids = [L2Level(price=Price(px), size=Size(sz)) for px, sz in bids[: self._depth]]
        self._asks = [L2Level(price=Price(px), size=Size(sz)) for px, sz in asks[: self._depth]]
        self._last_update_ts = ts
        self._version += 1
        self._snapshot = L2BookSnapshot(
            symbol=self.symbol,
            bids=tuple(self._bids),
            asks=tuple(self._asks),
            timestamp=ts,
            local_ts=received_at or datetime.now(UTC),
            version=self._version,
            connection_generation=connection_generation,
        )
        return self._snapshot

    def get_snapshot(self) -> L2BookSnapshot | None:
        """Return the snapshot created at update time without refreshing freshness."""
        return self._snapshot

    @property
    def best_bid(self) -> Price | None:
        """Best bid price."""
        return self._bids[0].price if self._bids else None

    @property
    def best_ask(self) -> Price | None:
        """Best ask price."""
        return self._asks[0].price if self._asks else None

    @property
    def mid_price(self) -> float | None:
        """Mid price = (best_bid + best_ask) / 2."""
        if self.best_bid is not None and self.best_ask is not None:
            return float((self.best_bid + self.best_ask) / 2)
        return None

    @property
    def spread(self) -> float | None:
        """Bid-ask spread."""
        if self.best_bid is not None and self.best_ask is not None:
            return float(self.best_ask - self.best_bid)
        return None

    @property
    def spread_bps(self) -> float | None:
        """Spread in basis points."""
        mid = self.mid_price
        sp = self.spread
        if mid and sp and mid > 0:
            return (sp / mid) * 10_000
        return None

    @property
    def last_update_ts(self) -> Timestamp | None:
        return self._last_update_ts

    def __repr__(self) -> str:
        bid = self.best_bid
        ask = self.best_ask
        return f"OrderBook({self.symbol}: bid={bid} ask={ask})"


class BookManager:
    """Manages order books for multiple symbols."""

    def __init__(self, depth: int = 20) -> None:
        self._depth = depth
        self._books: dict[Symbol, OrderBook] = {}

    def get_book(self, symbol: Symbol) -> OrderBook:
        """Get or create an order book for a symbol."""
        if symbol not in self._books:
            self._books[symbol] = OrderBook(symbol, self._depth)
        return self._books[symbol]

    def get_snapshot(self, symbol: Symbol) -> L2BookSnapshot | None:
        """Get a snapshot for a specific symbol."""
        book = self._books.get(symbol)
        return book.get_snapshot() if book else None

    def get_mid_price(self, symbol: Symbol) -> float | None:
        """Get mid price for a symbol."""
        book = self._books.get(symbol)
        return book.mid_price if book else None

    @property
    def symbols(self) -> list[Symbol]:
        """List of symbols with active books."""
        return list(self._books.keys())

    @property
    def active_books(self) -> Iterator[OrderBook]:
        """Iterate over all active books."""
        return iter(self._books.values())

File: crates/trading/src/market_data/book.rs
//! In-memory L2 order book manager, port of `src/hypeedge/market_data/book.py`.

use std::collections::HashMap;

use chrono::{DateTime, Utc};
use hypeedge_domain::decimal::{Decimal, Price, Size};
use hypeedge_domain::models::{L2BookSnapshot, L2Level};

/// Maintains an in-memory L2 order book for a single symbol. Updated from the
/// WebSocket feed; readable by strategies with zero latency.
pub struct OrderBook {
    symbol: String,
    depth: usize,
    bids: Vec<L2Level>,
    asks: Vec<L2Level>,
    last_update_ts: Option<i64>,
    version: u64,
    snapshot: Option<L2BookSnapshot>,
}

impl OrderBook {
    pub fn new(symbol: impl Into<String>, depth: usize) -> Self {
        Self {
            symbol: symbol.into(),
            depth,
            bids: Vec::new(),
            asks: Vec::new(),
            last_update_ts: None,
            version: 0,
            snapshot: None,
        }
    }

    /// Update the book with a full snapshot. `bids`/`asks` are `(price, size)`
    /// pairs best-first. Returns the created snapshot.
    pub fn update(
        &mut self,
        bids: &[(Price, Size)],
        asks: &[(Price, Size)],
        ts: i64,
        received_at: Option<DateTime<Utc>>,
        connection_generation: u32,
    ) -> L2BookSnapshot {
        self.bids = bids
            .iter()
            .take(self.depth)
            .map(|(p, s)| L2Level {
                price: *p,
                size: *s,
            })
            .collect();
        self.asks = asks
            .iter()
            .take(self.depth)
            .map(|(p, s)| L2Level {
                price: *p,
                size: *s,
            })
            .collect();
        self.last_update_ts = Some(ts);
        self.version += 1;
        self.snapshot = Some(L2BookSnapshot {
            symbol: self.symbol.clone(),
            bids: self.bids.clone(),
            asks: self.asks.clone(),
            timestamp: ts,
            local_ts: received_at.unwrap_or_else(Utc::now),
            version: self.version,
            connection_generation,
        });
        self.snapshot.clone().expect("just set")
    }

    pub fn get_snapshot(&self) -> Option<L2BookSnapshot> {
        self.snapshot.clone()
    }

    pub fn best_bid(&self) -> Option<Price> {
        self.bids.first().map(|l| l.price)
    }

    pub fn best_ask(&self) -> Option<Price> {
        self.asks.first().map(|l| l.price)
    }

    pub fn mid_price(&self) -> Option<f64> {
        let bid = self.best_bid()?;
        let ask = self.best_ask()?;
        Some(
            (bid.inner() + ask.inner())
                .div(Decimal::from_i128(2))
                .to_string()
                .parse()
                .unwrap_or(0.0),
        )
    }

    pub fn spread_bps(&self) -> Option<f64> {
        let mid = self.mid_price()?;
        let bid = self.best_bid()?;
        let ask = self.best_ask()?;
        if mid <= 0.0 {
            return None;
        }
        let spread =
            (ask.inner() - bid.inner()).div(Decimal::from_f64(mid).unwrap_or(Decimal::ONE));
        Some(spread.to_string().parse::<f64>().unwrap_or(0.0) * 10_000.0)
    }

    pub fn last_update_ts(&self) -> Option<i64> {
        self.last_update_ts
    }
}

/// Manages order books for multiple symbols.
pub struct BookManager {
    depth: usize,
    books: HashMap<String, OrderBook>,
}

impl Default for BookManager {
    fn default() -> Self {
        Self::new(20)
    }
}

impl BookManager {
    pub fn new(depth: usize) -> Self {
        Self {
            depth,
            books: HashMap::new(),
        }
    }

    pub fn get_book(&mut self, symbol: &str) -> &mut OrderBook {
        self.books
            .entry(symbol.to_string())
            .or_insert_with(|| OrderBook::new(symbol, self.depth))
    }

    pub fn get_snapshot(&self, symbol: &str) -> Option<L2BookSnapshot> {
        self.books.get(symbol).and_then(|b| b.get_snapshot())
    }

    /// Apply a normalized snapshot to the shared in-memory book.
    pub fn apply_snapshot(&mut self, snapshot: &L2BookSnapshot) -> L2BookSnapshot {
        let bids = snapshot
            .bids
            .iter()
            .map(|l| (l.price, l.size))
            .collect::<Vec<_>>();
        let asks = snapshot
            .asks
            .iter()
            .map(|l| (l.price, l.size))
            .collect::<Vec<_>>();
        self.get_book(&snapshot.symbol).update(
            &bids,
            &asks,
            snapshot.timestamp,
            Some(snapshot.local_ts),
            snapshot.connection_generation,
        )
    }

    pub fn get_mid_price(&self, symbol: &str) -> Option<f64> {
        self.books.get(symbol).and_then(|b| b.mid_price())
    }

    pub fn symbols(&self) -> Vec<String> {
        self.books.keys().cloned().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use hypeedge_domain::decimal::Decimal;

    fn px(s: &str) -> Price {
        Price::new(Decimal::from_str_strict(s).unwrap())
    }
    fn sz(s: &str) -> Size {
        Size::new(Decimal::from_str_strict(s).unwrap())
    }

    #[test]
    fn update_replaces_book_and_versions() {
        let mut book = OrderBook::new("BTC", 20);
        let s1 = book.update(
            &[(px("100"), sz("2"))],
            &[(px("101"), sz("3"))],
            1000,
            None,
            0,
        );
        assert_eq!(s1.version, 1);
        assert_eq!(book.best_bid(), Some(px("100")));
        assert_eq!(book.best_ask(), Some(px("101")));
        assert_eq!(s1.bids.len(), 1);

        let s2 = book.update(
            &[(px("99"), sz("5"))],
            &[(px("102"), sz("4"))],
            1001,
            None,
            1,
        );
        assert_eq!(s2.version, 2);
        assert_eq!(s2.connection_generation, 1);
        assert_eq!(book.mid_price().unwrap(), 100.5);
    }

    #[test]
    fn depth_limits_levels() {
        let mut book = OrderBook::new("ETH", 2);
        let bids = vec![(px("10"), sz("1")), (px("9"), sz("1")), (px("8"), sz("1"))];
        let s = book.update(&bids, &[], 1, None, 0);
        assert_eq!(s.bids.len(), 2, "only the best 2 levels retained");
        assert_eq!(s.bids[0].price, px("10"));
        assert_eq!(s.bids[1].price, px("9"));
    }

    #[test]
    fn book_manager_creates_and_reuses() {
        let mut mgr = BookManager::new(20);
        let snapshot = mgr.apply_snapshot(&L2BookSnapshot {
            symbol: "SOL".into(),
            bids: vec![L2Level {
                price: px("150"),
                size: sz("1"),
            }],
            asks: vec![L2Level {
                price: px("151"),
                size: sz("1"),
            }],
            timestamp: 5,
            local_ts: Utc::now(),
            version: 0,
            connection_generation: 0,
        });
        assert_eq!(
            snapshot.version, 1,
            "manager re-stamps version via the book"
        );
        assert_eq!(mgr.get_mid_price("SOL").unwrap(), 150.5);
        assert_eq!(mgr.symbols(), vec!["SOL".to_string()]);
    }
}
